latch_cli/services/sync.py

```diff
@@ -0,0 +1,328 @@
+        else {}
+    )
+
+    for name, (p, p_stat) in srcs.items():
+        is_dir = stat.S_ISDIR(p_stat.st_mode)
+
+        child = dest_children_by_name.get(name)
+        child_dest = f"{dest}/{name}"
+
+        skip = False
+        verb = "Uploading"
+        reason = "new"
+        if child is not None:
+            flt = child["finalLinkTarget"]
+            if flt["type"] == "DIR" and not is_dir:
+                # todo(maximsmol): confirm? pre-check?
+                click.secho(
+                    indent + f"`{dest}` is in the way of a file",
+                    fg="red",
+                )
+                continue
+
+            if flt["type"] != "DIR" and is_dir:
+                # todo(maximsmol): confirm? pre-check?
+                click.secho(
+                    indent + f"`{dest}` is in the way of a directory",
+                    fg="red",
+                )
+                continue
+
+            if flt["type"] == "OBJ":
+                meta = flt["ldataObjectMeta"]
+                remote_mtime = datetime.fromisoformat(meta["modifyTime"])
+
+                local_mtime = datetime.fromtimestamp(p_stat.st_mtime).astimezone()
+                if remote_mtime == local_mtime:
+                    verb = "Skipping"
+                    reason = "unmodified"
+                    skip = True
+                elif remote_mtime > local_mtime:
+                    verb = "Skipping"
+                    reason = "older"
+                    skip = True
+                else:
+                    verb = "Uploading"
+                    reason = "updated"
+            else:
+                reason = "existing"
+
+        if verb == "Uploading" and is_dir:
+            verb = "Syncing"
+
+        fg = "bright_blue"
+        dim = None
+        if verb == "Skipping":
+            fg = None
+            dim = True
+
+        click.echo(
+            click.style(
+                indent + verb + " ",
+                fg=fg,
+                dim=dim,
+            )
+            + click.style(
+                reason,
+                underline=True,
+                fg=fg,
+                dim=dim,
+            )
+            + click.style(
+                ": ",
+                fg=fg,
+                dim=dim,
+            )
+            + click.style(
+                str(p)
+                + ("" if not is_dir else "/")
+                + ("" if skip else click.style(" -> ", dim=True) + child_dest),
+                dim=dim,
+            )
+        )
+        if skip:
+            continue
+
+        if is_dir:
+            sub_srcs: Dict[str, Tuple[Path, os.stat_result]] = {}
+            for x in p.iterdir():
+                res = check_src(x, indent=indent + "  ")
+                if res is None:
+                    # todo(maximsmol): pre-check or confirm?
+                    continue
+
+                sub_srcs[x.name] = res
+            sync_rec(sub_srcs, child_dest, delete=delete, level=level + 1)
```

Comment:
ideally id like to avoid doing a gql request in every recursive call, but this is fine for now